quakesaver_client/fdsnws.py

```diff
@@ -0,0 +1,59 @@
+from __future__ import annotations
+
+import logging
+from datetime import datetime, timezone
+from pathlib import Path
+from typing import Literal, Optional
+
+import requests
+from pydantic import BaseModel, Field, PositiveFloat, constr
+
+from quakesaver_client import CorruptedDataError
+from quakesaver_client.util import assure_output_path
+
+NoData = Literal[204, 404]  # HTTP Error codes
+DataFormat = Literal["miniseed"]
+DataQuality = Literal["D", "R", "Q", "M", "B"]
+
+
+class FDSNWSDataselectQuery(BaseModel):
+    starttime: Optional[datetime] = None
+    endtime: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
+
+    # Add Set
+    # Reference http://www.fdsn.org/pdf/SEEDManual_V2.4.pdf [p37]
+    network: constr(max_length=2) = "*"
+    station: constr(max_length=5) = "*"
+    location: constr(max_length=2) = "*"
+    channel: constr(max_length=3) = "*"
+    quality: DataQuality = "B"
+
+    longestonly: bool = True
+    minimumlength: PositiveFloat = 0.0
+
+    format: DataFormat = "miniseed"
+    nodata: NoData = 204
+
+
+def dataselect(
+    uri: str,
+    params: FDSNWSDataselectQuery,
+    location_to_store: Path | str = None,
+) -> Path:
+    """Request FDSN waveform dat of the sensor."""
+    logging.debug("requesting waveform data for sensor %s.", uri)
+    location_to_store = assure_output_path(location_to_store)
+    response = requests.get(
+        url=f"{uri}/fdsnws/dataselect/1/query",
+        params=params.json(),
+    )
+
+    if response.status_code != 200:
+        raise CorruptedDataError(response.text)
+
+    filename = response.headers["Content-Disposition"].split("=")[1]
```

Comment:
Use `.get` on the dict to be more defensive.

Reply:
done